eval/src/vespa/eval/tensor/wrapped_simple_value.cpp

```diff
@@ -0,0 +1,260 @@
+            result.emplace(elem);
+        }
+    }
+    return result;
+}
+
+TensorSpec::Address
+extract_sparse_address(const TensorSpec::Address &address)
+{
+    TensorSpec::Address result;
+    for (const auto &elem : address) {
+        if (elem.second.is_mapped()) {
+            result.emplace(elem);
+        }
+    }
+    return result;
+}
+
+Tensor::UP
+maybe_wrap(Value::UP value)
+{
+    Tensor::UP retval;
+    if (auto tensor = dynamic_cast<Tensor *>(value.get())) {
+        retval.reset(tensor);
+        value.release();
+        return retval;
+    }
+    return std::make_unique<WrappedSimpleValue>(std::move(value));
+}
+
+struct PerformGenericApply {
+    template <typename CT>
+    static auto invoke(const Value &input, const CellFunction &func,
+                       const ValueBuilderFactory &factory)
+    {
+        const auto &type = input.type();
+        size_t subspace_size = type.dense_subspace_size();
+        size_t num_mapped = type.count_mapped_dimensions();
+        auto builder = factory.create_value_builder<CT>(type, num_mapped, subspace_size, input.index().size());
+        auto input_cells = input.cells().typify<CT>();
+        auto view = input.index().create_view({});
+        std::vector<vespalib::stringref> output_address(num_mapped);
+        std::vector<vespalib::stringref *> input_address;
+        for (auto & label : output_address) {
+            input_address.push_back(&label);
+        }
+        view->lookup({});
+        size_t subspace;
+        while (view->next_result(input_address, subspace)) {
+            auto dst = builder->add_subspace(output_address);
+            size_t input_offset = subspace_size * subspace;
+            for (size_t i = 0; i < subspace_size; ++i) {
+                dst[i] = func.apply(input_cells[input_offset + i]);
+            }
+        }
+        return builder->build(std::move(builder));
+    }
+};
+
+} // namespace <unnamed>
+
+
+eval::TensorSpec
+WrappedSimpleValue::toSpec() const
+{
+    return spec_from_value(_tensor);
+}
+
+void
+WrappedSimpleValue::accept(TensorVisitor &visitor) const
+{
+    TensorSpec myspec = toSpec();
+    TensorAddressBuilder addr;
+    for (const auto & cell : myspec.cells()) {
+        auto sparse_addr = sparsify_address(cell.first);
+        addr.clear();
+        for (const auto & dim_and_label : sparse_addr) {
+            addr.add(dim_and_label.first, dim_and_label.second.name);
+        }
+        visitor.visit(addr.build(), cell.second);
+    }
+}
+
+MemoryUsage
+WrappedSimpleValue::get_memory_usage() const
+{
+    size_t used = sizeof(WrappedSimpleValue);
+    if (_space) {
+        auto plus = _space->get_memory_usage();
+        plus.incUsedBytes(used);
+        plus.incAllocatedBytes(used);
+        return plus;
+    }
+    return MemoryUsage(used, used, 0, 0);
+}
+
+//-----------------------------------------------------------------------------
+
+Tensor::UP
+WrappedSimpleValue::apply(const CellFunction & func) const
```

Comment:
prefer a minal-aborting implementation over trying to fit the new world into the old-old-world by using the tensor::Tensor API.